contracts-bedrock: migrate to just (#11276)
ethereum-optimism/optimism: packages/contracts-bedrock/CONTRIBUTING.md

Our [Style Guide](STYLE_GUIDE.md) contains information about the project structure, syntax preferences, naming conventions, and more. Please take a look at it before submitting a PR, and let us know if you spot inconsistencies!

> Once you've read the styleguide and are ready to work on your PR, there are a plethora of useful `just` scripts to know about that will help you with development:

Reviewer: nit, feel free to tweak the phrasing, you get the idea
```suggestion
Once you've read the style guide and are ready to work on your PR, there are a plethora of useful `just` scripts to know about that will help you with development.
You can run `just -l` to list them all, some of the key ones are:
```

1. `just build` Builds the smart contracts.
1. `just test` Runs the full `forge` test suite.
1  `just gas-snapshot` Generates the gas snapshot for the smart contracts.
1. `just semver-lock` Generates the semver lockfile.

> 1. `just storage-snapshot` Generates the storage lockfile.

Reviewer: looks like this one was outdated
```suggestion
1. `just snapshots` Generates the storage and ABI snapshots.
```

1. `just autogen-invariant-docs` Generates the invariant test documentation.
1. `just clean` Removes all build artifacts for `forge` and `go` compilations.
1. `just validate-spacers` Validates the positions of the storage slot spacers.
1. `just validate-deploy-configs` Validates the deployment configurations in `deploy-config`

> 1. `just slither` Runs the slither static analysis tool on the smart contracts.

Reviewer: looks like we never had a command in the package.json to run slither either, though there are instructions for it in the readme
```suggestion
```

In order to make sure that we don't accidentally overwrite storage slots, contract storage layouts are checked to make sure spacing is correct.

> This uses the `.storage-layout` file to check contract spacing. Run `just validate-spacers` to check the spacing of all contracts.

Reviewer: I think this is outdated too and we don't use `.storage-layout` right? I know the text wasn't changed in this PR, but since we're modifying it here anyway it would be good to fix

Author: I resolved this

#### Gas Snapshots

> We use forge's `gas-snapshot` subcommand to produce a gas snapshot for most tests within our suite. CI will check that the gas snapshot has been updated properly when it runs, so make sure to run `just gas-snapshot`!

Reviewer 2: I know this is outside the scope of the PR, but good to use this opportunity to keep it updated.

```suggestion
We use forge's `gas-snapshot` subcommand to produce a gas snapshot for tests in `Benchmark.t.sol`. CI will check that the gas snapshot has been updated properly when it runs, so make sure to run `just gas-snapshot`!
```

#### Semver Locking

Many of our smart contracts are semantically versioned. To make sure that changes are not made to a contract without deliberately bumping its version, we commit to the source code and the creation bytecode of its dependencies in a lockfile. Consult the [Style Guide](./STYLE_GUIDE.md#Versioning) for more information about how our contracts are versioned.

#### Storage Snapshots

> Due to the many proxied contracts in Optimism's protocol, we automate tracking the diff to storage layouts of the contracts in the project. This is to ensure that we don't break a proxy by upgrading its implementation to a contract with a different storage layout. To generate the storage lockfile, run `just storage-snapshot`.

Reviewer: ```suggestion
Due to the many proxied contracts in Optimism's protocol, we automate tracking the diff to storage layouts of the contracts in the project. This is to ensure that we don't break a proxy by upgrading its implementation to a contract with a different storage layout. To generate the storage lockfile, run `just snapshots`.
```
